src/main/scala/sparkz/core/network/NetworkController.scala

```diff
@@ -195,8 +191,22 @@ class NetworkController(settings: NetworkSettings,
       connectionToPeer(activeConnections, unconfirmedConnections)
   }
 
+  private def handleNewConnectionCreation(remoteAddress: InetSocketAddress, localAddress: InetSocketAddress): Unit = {
+    val connectionDirection: ConnectionDirection = if (unconfirmedConnections.contains(remoteAddress)) Outgoing else Incoming
+    val connectionId = ConnectionId(remoteAddress, localAddress, connectionDirection)
+
+    log.info(s"Unconfirmed connection: ($remoteAddress, $localAddress) => $connectionId")
+
+    if (connectionDirection.isOutgoing && canEstablishNewOutgoingConnection)
+      createPeerConnectionHandler(connectionId, sender())
+    else if (connectionDirection.isIncoming && canEstablishNewIncomingConnection)
+      peerManagerRef ! ConfirmConnection(connectionId, sender())
+    else
+      self ! ConnectionDenied(connectionId, sender())
```

Comment:
can it happen that we get to this case because we reach the limit of outgoing connection?
If yes, then look at the log line inside `ConnectionDenied`. At the moment it always print `log.info(s"Incoming connection from ${connectionId.remoteAddress} denied")`. Must be adopted to the new behavior (if it is the case)

Reply:
You're right, that message doesn't make sense anymore with the new logic. Gonna change it 👍